external-plugins/phased/plugin/handler/handler.go

```diff
@@ -0,0 +1,367 @@
+}
+
+type GitHubEvent struct {
+	Type    string
+	GUID    string
+	Payload []byte
+}
+
+type githubClientInterface interface {
+	GetPullRequest(string, string, int) (*github.PullRequest, error)
+	CreateComment(org, repo string, number int, comment string) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+}
+
+type loadConfigBytesFunc func(h *GitHubEventsHandler, org, repo string) ([]byte, []byte, error)
+
+var LoadConfigBytesFunc loadConfigBytesFunc = loadConfigBytes
+
+type GitHubEventsHandler struct {
+	eventsChan       <-chan *GitHubEvent
+	logger           *logrus.Logger
+	ghClient         githubClientInterface
+	gitClientFactory gitv2.ClientFactory
+	prowConfigPath   string
+	jobsConfigBase   string
+	prowLocation     string
+}
+
+func NewGitHubEventsHandler(
+	eventsChan <-chan *GitHubEvent,
+	logger *logrus.Logger,
+	ghClient githubClientInterface,
+	prowConfigPath string,
+	jobsConfigBase string,
+	prowLocation string,
+	gitClientFactory gitv2.ClientFactory) *GitHubEventsHandler {
+
+	return &GitHubEventsHandler{
+		eventsChan:       eventsChan,
+		logger:           logger,
+		ghClient:         ghClient,
+		prowConfigPath:   prowConfigPath,
+		jobsConfigBase:   jobsConfigBase,
+		prowLocation:     prowLocation,
+		gitClientFactory: gitClientFactory,
+	}
+}
+
+func (h *GitHubEventsHandler) Handle(incomingEvent *GitHubEvent) {
+	log.Infoln("GitHub events handler started")
+	eventLog := log.WithField("event-guid", incomingEvent.GUID)
+	switch incomingEvent.Type {
+	case "pull_request":
+		eventLog.Infoln("Handling pull request event")
+		var event github.PullRequestEvent
+		if err := json.Unmarshal(incomingEvent.Payload, &event); err != nil {
+			eventLog.WithError(err).Error("Could not unmarshal event.")
+			return
+		}
+		h.handlePullRequestEvent(eventLog, &event)
+	default:
+		log.Infoln("Dropping irrelevant:", incomingEvent.Type, incomingEvent.GUID)
+	}
+}
+
+// For unit tests, as we create a local git NewFakeClient
+func (h *GitHubEventsHandler) SetLocalConfLoad() {
+	LoadConfigBytesFunc = loadLocalConfigBytes
+}
+
+func (h *GitHubEventsHandler) handlePullRequestEvent(log *logrus.Entry, event *github.PullRequestEvent) {
+	log.Infof("Handling updated pull request: %s [%d]", event.Repo.FullName, event.PullRequest.Number)
+
+	if !h.shouldActOnPREvent(event) {
+		return
+	}
+
+	org, repo, err := gitv2.OrgRepo(event.Repo.FullName)
+	if err != nil {
+		log.WithError(err).Errorf("Could not get org/repo from the event")
+		return
+	}
+
+	shouldRun, err := h.shouldRunPhase2(org, repo, event.Label.Name, event.PullRequest.Number)
+	if err != nil || !shouldRun {
+		return
+	}
+
+	pr, err := h.ghClient.GetPullRequest(org, repo, event.PullRequest.Number)
+	if err != nil {
+		log.WithError(err).Errorf("Could not get PR number %d", event.PullRequest.Number)
+		return
+	}
+
+	presumbits, err := h.loadPresubmits(*pr)
```

Comment:
It would be more efficient to load the config on plugin startup and restart plugin when configmap changes. But that can be done in a follow up to avoid premature optimization.

Reply:
we cant restart the plugin, it is prone for races, when the plugin is down, events can be lost 
(assuming the plugin doesnt get all the history events)

anyhow we can do a follow up optimization that checks if the file was changed indeed (according git log etc),
and load only in this case, the files doesn't change a lot